autonomy/cli/scaffold_fsm.py

```diff
@@ -0,0 +1,769 @@
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""
+Implement a scaffold sub-command to scaffold ABCI skills.
+
+This module patches the 'aea scaffold' command so to add a new subcommand for scaffolding a skill
+ starting from FSM specification.
+"""
+import re
+import shutil
+from abc import ABC, abstractmethod
+from pathlib import Path
+from textwrap import dedent, indent
+from typing import Dict, Set
+
+import click
+from aea.cli.add import add_item
+from aea.cli.fingerprint import fingerprint_item
+from aea.cli.scaffold import scaffold, scaffold_item
+from aea.cli.utils.context import Context
+from aea.cli.utils.decorators import pass_ctx
+from aea.configurations.base import (
+    AgentConfig,
+    SkillComponentConfiguration,
+    SkillConfig,
+)
+from aea.configurations.constants import (
+    DEFAULT_AEA_CONFIG_FILE,
+    DEFAULT_SKILL_CONFIG_FILE,
+    SKILL,
+    SKILLS,
+)
+
+# the decoration does side-effect on the 'aea scaffold' command
+from aea.configurations.data_types import CRUDCollection, PublicId
+
+from autonomy.analyse.abci.app_spec import DFA
+
+
+FILE_HEADER = """\
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2022 Valory AG
+#   Copyright 2018-2021 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""
+
+ROUNDS_FILENAME = "rounds.py"
+BEHAVIOURS_FILENAME = "behaviours.py"
+MODELS_FILENAME = "models.py"
+HANDLERS_FILENAME = "handlers.py"
+
+DEGENERATE_ROUND = "DegenerateRound"
+ABSTRACT_ROUND = "AbstractRound"
+
+
+def _remove_quotes(input_str: str) -> str:
+    """Remove single or double quotes from a string."""
+    return input_str.replace("'", "").replace('"', "")
+
+
+def _get_abci_app_cls_name_from_dfa(dfa: DFA) -> str:
+    """Get the Abci app class name from a DFA object."""
+    return dfa.label.split(".")[-1]
+
+
+def _try_get_behaviour_cls_name_from_round_cls_name(round_cls_name: str) -> str:
+    """
+    Try to get the behaviour class name from the round class name.
+
+    It tries to replace the suffix "Round" with "Behaviour".
+
+    :param round_cls_name: the round class name
+    :return: the new behaviour class name
+    """
+    return re.sub("(.*)Round", "\\1Behaviour", round_cls_name)
```

Comment:
`_try_get_behaviour_cls_name_from_round_cls_name` and `_try_get_round_behaviour_cls_name_from_abci_app_cls_name` are supposed to become design constraints?

Reply:
Yes; we need to impose some constraints on spec naming conventions. Probably good to enforce them explicitly on the spec model side.

Reply:
https://github.com/valory-xyz/open-autonomy/issues/1196